### ISSUE TITLE
LockWaitTimeoutException in CleanupHelper Causing API Failures

### ISSUE THREAD

**Issue by Digi92** [mosparo/mosparo#275]
Hi,
we have noticed a problem with the [cleanupHelper->cleanup](https://github.com/mosparo/mosparo/blob/da2aa17765fbfcaa4ef11ec3cbafeea4a9d9fbaa/src/Helper/CleanupHelper.php#L29) over the last few days.

This has caused this error "`Uncaught PHP Exception Doctrine\DBAL\Exception\LockWaitTimeoutException: "An exception occurred while executing a query: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction" at ExceptionConverter.php line 37`" several times.

When this function was called, the tables “submission” and/or “submit_token” were in a `locked` state for a long time. 
This `locked` state caused all API requests to fail and users were unable to validate forms. 

After some debugging I was able to determine that it is triggered by this query: 
[`DELETE Mosparo\Entity\Submission s
                    WHERE (SELECT COUNT(st.id) FROM Mosparo\Entity\SubmitToken st WHERE st.id = s.submitToken OR st.lastSubmission = s.id) = 0`](https://github.com/mosparo/mosparo/blob/da2aa17765fbfcaa4ef11ec3cbafeea4a9d9fbaa/src/Helper/CleanupHelper.php#L168)

Due to the fact that the select is a subquery of DELETE, the table is locked while the read operation is executed instead of being writable.

Splitting these two queries fixed the error for us on Live.
I have improved the performance of the select.
```
$query = $this->entityManager->createQuery('
    SELECT s.id
    FROM Mosparo\Entity\Submission s
    WHERE NOT EXISTS (
        SELECT 1 FROM Mosparo\Entity\SubmitToken st
        WHERE st.id = s.submitToken OR st.lastSubmission = s.id
    )
');
$submissionIds = $query->getResult();

$batchSize = 50;
for ($i = 0; $i < count($submissionIds); $i += $batchSize) {
    $batchIds = array_slice($submissionIds, $i, $batchSize);

    $deleteQuery = $this->entityManager->createQuery('
        DELETE Mosparo\Entity\Submission s
        WHERE s.id IN (:ids)
    ');
    $deleteQuery->setParameter('ids', $batchIds);
    $deleteQuery->execute();
}
unset($query);
```

What do you think about the problem and my solution?
Do you have a better suggestion?

My setup:
Version: v1.2.5
PHP: 8.3
SQL: 10.6-MariaDB

We had a promotion in the last few days which resulted in significantly more entries in the tables than before.
I estimated that this caused the query to take noticeably longer.

The table “submit_token” has approx. 41000 entries and the table “submission” has 38000 entries.

Kind regards,
Digi92

**Comment by zepich** [comment]
Hi @Digi92 

Thank you very much for your detailed report and the fix for this problem.

We switched to the subquery method last year when optimizing the cleanup logic. The optimization back then was to prevent issues with running out of memory (#148). With this subquery, we do not have to care about the memory issue—at least not in PHP. The query that generated the problem for you was added later, but we used the same style as for the other two delete queries.

Of course, using a subquery will lock the table and, therefore, generate completely different problems — as you noticed.

Your solution is almost perfect! However, it introduces another potential memory issue since we do not know how large `$submissionIds` will be, and it could require too much memory.

Because of that, I've adjusted your solution to this one:

```php
// The limiter prevents possible endless loops.
for ($limiter = 0; $limiter < 1000; $limiter++) {
    $query = $this->entityManager->createQuery('
            SELECT s.id
            FROM Mosparo\Entity\Submission s
            WHERE NOT EXISTS (
                SELECT 1 FROM Mosparo\Entity\SubmitToken st
                WHERE st.id = s.submitToken OR st.lastSubmission = s.id
            )
        ')
        ->setMaxResults(50);
    $submissionIds = $query->getResult();
    if (!$submissionIds) {
        break;
    }

    $deleteQuery = $this->entityManager->createQuery('
            DELETE Mosparo\Entity\Submission s
            WHERE s.id IN (:ids)
        ')
        ->setParameter('ids', $submissionIds);
    $deleteQuery->execute();
    unset($deleteQuery);
}

unset($query);
unset($submissionIds);
```

Since we're limiting the first `SELECT`, we have a maximum of 50 IDs in the array, so there should be no memory issue. Also, the locking is not a problem because we only try to delete these 50 submission IDs that we received from the `SELECT` (as in your solution). 

The condition in the `for` is only there so as not to run in an endless loop. Of course, it would not be an infinite loop, but it would run very long in the worst case if there were a lot of such 'broken' submissions.

What do you think about this approach?

Kind regards,
zepich

**Comment by Digi92** [comment]
Hi @zepich,
it looks good to me.

Additionally, now that I think about it, maybe we should also add notice logging that this cleanup command was started and finished.
When debugging, it took me a bit to realize that the cleanup was causing this problem.

Kind regards,
Digi92

**Comment by zepich** [comment]
Hi @Digi92 

We've just released v1.2.7, which contains the fix for the cleanup process.

I'm closing this issue now. Thank you again for your help, and have a great day!

Kind regards,

zepich
